src/client/mod.rs

```diff
@@ -216,12 +260,44 @@ impl Client {
         })
     }
 
-    pub fn begin(&mut self, n_metrics: u64) -> io::Result<&mut Client> {
+    /// Begins registration of singleton metrics only
+    pub fn begin_metrics(&mut self, n_metrics: u64) -> io::Result<&mut Client> {
+        self.begin_all(0, 0, 0, n_metrics)
+    }
+
+    /// Begins registration of singleton and instance metrics
+    /// - `n_indoms` is the number of unique indoms
+    /// - `n_instances` is the number of unique instances
+    pub fn begin_all(&mut self, n_indoms: u64, n_instances: u64, n_instance_metrics: u64, n_metrics: u64) -> io::Result<&mut Client> {
         self.wi.n_metrics = n_metrics;
+        self.wi.n_toc = 2; // metric + value
+
+        if n_indoms > 0 && n_instances > 0 && n_instance_metrics > 0 {
+            self.wi.n_indoms = n_indoms;
+            self.wi.n_instances = n_instances;
+            self.wi.n_instance_metrics = n_instance_metrics;
+            self.wi.n_toc += 2; // indom + instance
+        }
+        self.wi.n_metric_blks = self.wi.n_metrics + self.wi.n_instance_metrics;
+        self.wi.n_value_blks =
+             self.wi.n_metrics +
+             self.wi.n_instances * self.wi.n_instance_metrics; // FIXME: very wasteful allocation in some cases
```

Comment:
For the exporting all metrics in one function we'll need common trait that they all implement. I think it would be better to design that trait after we implement the higher-level metrics as well.